rfcs/0000-graphql-api.md

```diff
@@ -0,0 +1,411 @@
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- The payment and block history for the wallets could be only stored on the client. This would
+  make the node only be responsible for storing the private keys of the wallets. This results in a somewhat
+  simpler/lighter full node, with the obvious cost that whenever you close your wallet app, you're missing
+  any transactions that might be happening, which isn't great for the wallet experience.
+- The wallet could be responsible for storing the private keys as well, but then the node would have to ask the wallet
+  for them whenever it needed them for snarking etc. Probably not what we want.
+
+Why GraphQL?
+- Defines a typed interface for communication between client and server.
+- Interface is explorable/discoverable with great existing tools, making development easier.
+- Strong OCaml/Reason support without requiring that consumers of the api be written in a specific language.
+
+Alternatives to GraphQL:
+- Falcor (or similar)
+- Completely custom alternative.
+
+We considered a custom alternative as a single wallet app connected to a single node instance isn't a very standard
+setup for graphql app, which we saw as more associated with public web apis with many consumers. However, all the work
+in this case would obviously have to be from scratch, representing potentially significant time investment to get stable
+for questionable gain. Client libraries would also have to be written for any languages that wanted to interface, whereas
+GraphQL already supports many and seems to have pretty good community buy-in.
+
+## Prior art
+[prior-art]: #prior-art
+
+- RPC interface that the commandline client uses: This has inspired the graphql api though in some cases we have tried to
+simplify the interface and avoid any binary serialization that might rely on having ocaml at both ends.
+- REST server in the node: A simple interface that uses rest to deliver some status pages.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Should we be able to query the node for the amount of the balance that corresponds to unknown/missed transactions?
+  We could also calculate this client-side by summing up all the payments/fees/coinbases and comparing to the balance
+  queried from the node.
+- Several apis involve objects that get incrementally updated during consensus
+  It's important that we have a reliable way to associate updates with their
+  corresponding objects in memory. We don't have a concept of IDs for most of these
+  objects, but if we could create one by hashing together enough properties of the 
+  objects, it might help with this and avoid accidents comparing fields in an insufficient way.
+- Authentication
+- Potentially out of scope: How will this evolve and be used in the future, when most wallets will just run a light node locally?
```

Comment:
I'd think we will likely always want an api for accessing full nodes (in many cases remotely) even when we do have light nodes embedded in the wallet. I wouldn't be surprised if even in the case where you have the option of a light node locally while being connected to a full node, you would just rely on the full node (since you wouldn't need to make queries to the network about the current state of the ledger etc).

On the webworker question, it seems possible, if not common. Apollo exposes something called "link" which lets you tell the client how to resolve the queries. [Someone](https://www.youtube.com/watch?v=YOCrJ5vRCnw) used this to build a [prototype](https://www.npmjs.com/package/apollo-link-webworker) that talks to a graphql server in a webworker. There's also stuff like [graphql-anywhere](https://github.com/apollographql/apollo-client/tree/master/packages/graphql-anywhere), which could be even more flexible.

The less explored territory is probably on the "server side", where we might need to write a simplified custom backend (in addition to the existing async and lwt ones) for ocaml-graphql-server that communicates within the browser. Of course, we would probably have to do something similar to this even if we didn't use graphql for in-browser communication.